Internal Logging

MadVikingGod opened the pull request:
# What is this?
This is an experiment to enable logging from the API and SDK. 

It creates a simple user facing interface to set a logger, and then exposes that `logr.Logger` in the `internal/debug` package.  

# What is not included in this PR?
- Instrumenting some/most/all of the API, SDK, or exporters.  I wanted to explore how simple of an interface I could make, and that can come in a later patch set.
- Guidelines on verbosity.  I have included a simple blurb, but it is by no means fully flushed out. 
- How to warn users that using this with a logger that exports to the otel log means messages might be lost if incorrectly setup.

# How do you use it?
The simplest way is to run the NamedTracer example.  This should output to stderr the only line that was instrumented in the project.
```
cd example/namedtracer
go run . >/dev/null
```

Closes #1068

pellared commented on debug/debug.go:
```diff
@@ -0,0 +1,36 @@
+)
+
+// SetLogger configures the logger used internally to opentelemetry.
+func SetLogger(logger logr.Logger) {
```
What do you think of changing the `logger` type to an interface which would be "compatible" with `logr`?

Pros:
- we could remove the coupling to `logr`. The user could inject its own logger created using `logr` or use its own logger with API similar to `logr`.
- we always have the possibility to create our own implementation with better performance than `logr`

Cons:
- can lead to worse performance b/c of `interface{}` usage (would need to be measured if it is really a problem) [more](https://github.com/go-logr/logr#why-is-this-not-a-pure-interface)

### Background

Here is a great article by @nblumhardt where he describes similar problems and approaches for .NET: https://nblumhardt.com/2016/04/which-logging-abstraction-should-i-use/

MadVikingGod replied:
While I would like to, I can't actually make an interface that is 100% decoupled and uses all of the features of the logr logger.

For example I can't have the [WithValues()](https://pkg.go.dev/github.com/go-logr/logr#Logger.WithValues).  Because they return a `logr.Logger`, and we couldn't we would just return our internal Logger (`debug.Logger`).  

This would then limit this interface to only the `Error()` and `Info()` methods.

We could of course build our own internal logger struct, and have an interface for the logr to our struct, but that would make it that much more awkward to use something like `zap` or `loggrus` because our users would have to wrap those loggers in logr and then in our struct.

pellared replied:
> While I would like to, I can't actually make an interface that is 100% decoupled and uses all of the features of the logr logger. For example I can't have the WithValues(). Because they return a logr.Logger, and we couldn't we would just return our internal Logger (debug.Logger).

True. But do we want all want the whole API and all of the features of `logr`? I would rather prefer to have a minimal API surface. E.g. hardcode the logging levels and just create an example adapter for the `logr` (e.g. in contrib repo) so that we get other adapters out-of-the-box with minimal effort.

> We could of course build our own internal logger struct, and have an interface for the logr to our struct, but that would make it that much more awkward to use something like zap or loggrus because our users would have to wrap those loggers in logr and then in our struct.

Yeah, true... That is what I meant when I was writing "For users using it, adding an interface is just an unnecessary layer of abstraction.". Still, I think it might be worth comparing and experimenting with.

_I will try to create a PR but I just need to find some time to do it..._

MadVikingGod replied:
Going to try and cover a few topics here:

**Size of the `logr` API**
Right now the `logr` API is about twice as big as I see us using, with `Error`, `Info`, `V`, `WithValues`, and maybe `WithName` being the biggest candidate.  That being said we only over expose this API internally, so we can change it internally if we need to. The customer facing API consists of just a SetLogger.  While we can't update this if we did want to move to an internal implementation we could write an adapter to the new format, and expose a new API that allows directly setting the new logger.

**Internal implementation**
The biggest downside I see to us implementing our own internal implementation is setting up logging goes from:
```
zapLog, _:= zap.NewDevelopment()
debug.SetLogger(zapr.NewLogger(zapLog))
```
to
```
zapLog, _:= zap.NewDevelopment()
zaprLog := zapr.NewLogger(zapLog)
debug.SetLogger(otelLog.NewLogger(zaprLog))
```
This is additional complexity that doesn't seem to gain us much.  The current plan would just to have a transition layer from our internal semantics of levels to the appropiate V(lvl) within logr.

**Conventions**
I was going to try and punt on this as much as possible, but lets hash a bit of this out.  You asked in a different thread about enforcing conventions, eg calling V(100).  First off I think we need to try to instrument our code more before we set hard conventions for new code and think about enforcement of them.  I can already see a need for three different logging levels e.g. 

- Info, things that would print rarely to inform user of the general configuration/state of the system config would be a good candidate
- verbose, things that would readily inform the user of ongoing operations of system, like a report when a batcher sent a batch
- debug, extra details of EVERYTHING that is happening within the system.

The main problem with this is that even within the `*r` world there isn't a convention of what and how `V(lvl)` is represented within the actual loggers (`zap`, `loggrus`, `zerolog`).  What this means is that how to enable different verbosity levels between loggers, and what the convention is in that logger is different.  My suggestion to solve this is to build examples with two or three different loggers to give users an easy onboarding experience.

MadVikingGod replied:
An update on this thread:

** Size of the `logr` API **
I unexported the interal Log, and reexposed it interally via `Info`, `Error`, and `Debug`.  This can allow us to control usage through out the project.  It will not decouple us for `logr`.  We save a lot of complexity by not creating our own interface, and mainting bridges to other logging libraries.

** Conventions **
I addressed it in the main thread, but the TLDR is some exploratory work gave me confidance that there is some uniformity in the logging space, and we can collapse down to three levels.

pellared replied:
I like it as it decreases the coupling and increases maintainablilty.

pellared commented on internal/debug/debug.go:
```diff
@@ -0,0 +1,29 @@
+	"github.com/go-logr/logr"
+)
+
+var Log logr.Logger = logr.Discard()
```
Can we unexport this global variable and add a `SetLogger` function to be able to set it?

This way we can make sure that in other packages we can use only functions exposed by this package instead of everything that `logr` exposes (e.g. use some magic logging level like in `sdk/trace/batch_span_processor.go`😄 )

MadVikingGod replied:
I think that exposing the V() and WithValues() methods internally are the compelling features of this approach.

MadVikingGod replied:
Unexported, and exposed a smaller API

pellared commented on internal/debug/debug.go:
```diff
@@ -0,0 +1,29 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package debug // import "go.opentelemetry.io/otel/internal/debug"
```
How about renaming to `log`? 😉

MadVikingGod replied:
I didn't want to squat on the log namespace because we will be adding a whole logging section in the near future.

MadVikingGod commented on debug/debug.go:
```diff
@@ -0,0 +1,36 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package debug // import "go.opentelemetry.io/otel/debug"
```
Could they, yes.  I opted for a package of some kind to hide the API from the user which should only need this if something is going wrong.  Do you have a suggestion on what might be a more appropriate name?

MadVikingGod replied:
Well if we go and instrument our API or  SDK with this there would be no way to have this as a module, and optionally include it.

What would be a good name for the setting of this to convey what it does?  I'm concerned that going with `otel.SetLogger` or even `otel.SetInternalLogger` isn't enough to convey that this doesn't expose the OpenTelemetry (non-existant) API for logging.  Keeping it separate, by a namespace of some kind, is a signal that this does something different from the SetTracerProvider etc.

MrAlias replied:
> What would be a good name for the setting of this to convey what it does? I'm concerned that going with otel.SetLogger or even otel.SetInternalLogger isn't enough to convey that this doesn't expose the OpenTelemetry (non-existant) API for logging. Keeping it separate, by a namespace of some kind, is a signal that this does something different from the SetTracerProvider etc.

Yeah, that is a good problem to recognize. Honestly, I'm not sure outside of a good name there is going to be a way to communicate what your are setting without being able to rely on the reader reading the Go docs. Even if it is in its own package.

Would it be worth considering names like `MsgHandler` or `MessageHandler`?

MadVikingGod replied:
In the current iteration it is `otel.SetLogger()`.  While I'm not happy with it, I am willing to live with it, only because we will most likely call any logging API something akin to `SetLoggerProvider`.

Are we willing to live with this, or would you like to see an iteration with MessageHandler?

Aneurysm9 replied:
I have a slight preference for `otel.SetInternalLogger()`, but I think either is workable.

MadVikingGod commented on debug/debug.go:
```diff
@@ -0,0 +1,36 @@
+}
+
+// SetDefaultLogger configures the internal logger to use stderr and show verbose logging messages.
+func SetDefaultLogger() {
```
You are right, the default is a discard.  I wanted a name that someone stepping into this realm would recognize as what to use to start getting logging out of their instrumentation, any suggestions?

MadVikingGod replied:
I changed the default logger to `stdr`.  Because I unexported the globalLogger I also removed the need for Higher verbosity levels.

pellared commented on sdk/trace/batch_span_processor.go:
```diff
@@ -216,6 +218,7 @@ func (bsp *batchSpanProcessor) exportSpans(ctx context.Context) error {
 	}
 
 	if l := len(bsp.batch); l > 0 {
+		debug.Log.V(5).Info("exporting spans", "count", len(bsp.batch))
```
Do we want to use `logr` API directly like here? The API surface is quite large. How do we ensure that we follow some conventions? Do we want to create some abstraction layer to not use `logr` directly? Do we want to create our own linter to e.g. not allow `V(100)` or is it not a problem?

MadVikingGod replied:
I answered it [above](https://github.com/open-telemetry/opentelemetry-go/pull/2343#discussion_r746071510).  TLDR:
We don't have a convention, and we probably won't until we instrument things.  When we get to that point we can look at how to enforce it.

MadVikingGod replied:
Answered in recent [Update](https://github.com/open-telemetry/opentelemetry-go/pull/2343#issuecomment-971775390).  

TLDR:
Did a survey of the `logr` ecosystem, and there is a simple convention that came out of it.

MadVikingGod commented:
So I was not aware of the `otel.ErrorHandler` when this was written.  I'm working on a patch that will have these parts work in unison so we don't have to duplicate efforts in both systems.  My first pass will have the default errorHandler be a function that calls the internal `logger.Error`, with a message of the `err.Error()`

MadVikingGod commented:
# Update
This now includes a rework of the error handeling. I'm not a fan of the scope creep, but I felt it necessary to have these interfaces work together as much as possible in unison.  I simplified how the error handler is managed, and changed it's default to use the internal logger.  I also made the default internal logger print Error messages to stderr, this perserves the behavior of the current errorHandler.

# Conventions
When I first started on this because each logging implementation uses different levels internally (logrus uses 1 - -5, but glog, klog, and zap use 1 - 5 etc), I didn't know if there would be an easy convention we could set that would be easy to explain to users.  To drive this I experimented with all of the loggers [here](https://github.com/MadVikingGod/logr-survey), and found that MOST logr implementations use 0 as their default, and allow all Error messages.  This makes setting a convention easy.  To that extent what I set was three different levels Error, Info (lvl 1 off by default for most loggers), and debug (lvl 5 off on all loggers).  I just need a bit of help writing the SetLogger documentation to explain how to see the different levels.

pellared commented on internal/debug/debug.go:
```diff
@@ -0,0 +1,53 @@
+}
+
+// Error prints messages about exceptional states of the API or SDK.
+func Error(err error, msg string, keysAndValues ...interface{}) {
```
Do we intend to log transient failures? If so then on which level? I tend to use `Warn` for such cases

MadVikingGod replied:
So I think it would depend.  My general idea with this is if someone were to turn on V(1), they would get something like 5-10 messages a minute or so.  This number was picked from thin air to represent a low but manageable number of messages.  Things that describe the current state of the SDK.  
If your transient failure was something that needed immediate attention then it would be in Error, which all users would see.  If not then it would be in Info, which describes the last known state of the SDK, and some attempt should be made to no have:
```
10:01:04.05  Failed to connect
10:01:04.06  Failed to connect
10:01:04.07  Failed to connect
10:01:04.08  Failed to connect
10:01:04.09  Failed to connect
```

pellared replied:
👍  resolving

pellared commented on internal/debug/doc.go:
```diff
@@ -0,0 +1,18 @@
+// limitations under the License.
+
+/*
+Package debug provides access to enable internal debugging of opentelemetry.
```
should this package be internal? don't we want to use it e.g. in go-contrib?

MadVikingGod replied:
Not yet.  That would significantly expand the scope of this PR.  The idea is something that can be used internally by the SDK.  If we make this an exposed API for contrib, then we also expose it to end-users, which brings up a whole different set of concerns.  Things like:

- Should an instrument wrapper use the SDK internal error handling or the library's error handling?
- Would library writers adopt this logging system when we will want to push to use the "(un)official" logging API that goes through the SDK's processing steps?
- Is this API robust enough to cover most users' use cases? (my guess is no)

In the end, the only misses that are glaring are any contrib exporters.

MrAlias commented on handler.go:
```diff
@@ -15,90 +15,66 @@
-	// delegateErrorHandlerOnce ensures that a user provided ErrorHandler is
-	// only ever registered once.
-	delegateErrorHandlerOnce sync.Once
+	globalErrorHandler = &errorHandlerDelegate{
```
Can we remove the changes to the ErrorHandler for this PR?

I'm not entirely convinced this is a good idea (what does it mean to set a logger versus an error handler), but I also don't think my uncertainty should block the progress of adding internal logging. If we can isolate the changes being made in this PR to just adding internal logging it would be easier to get this merged IMO.

MrAlias replied:
I also get that it might not be correct to abandon this unification without it being recognized. If that is the case can we create an issue to track the work instead?

MadVikingGod replied:
On second thought the changes to how error handling is delegated are a separate issue.  I originally included it because it was the exact same thing I was going to use to delegate the logger before I realized I don't needed.  This commit will pull those out, and I will open a separate PR.

We should still use the internal logging as the default, this is a much smaller change.  The reasonfor it is what would I expect from changing the logging of the SDK.  I would think if I changed it to a pretty formatter (or json or anything really), I would expect that ALL logs, including those from the contrib libraries usingthe error handler, would start to use the new formater.  I would be surprised, in a bad way, to find out that to get the contrib errors to be formated I would also have to write an error handler.

The behavior that I would expect:
The default is internal logger, so if you change the logger everything from SDK/API and contrib gets the new formatting
The Error handler would be more specific, so if you change that it won't affect the more verbose logs, but it would change Errors from the SDK and contrib.

pellared replied:
Resolving per https://github.com/open-telemetry/opentelemetry-go/issues/2417

MadVikingGod commented on internal/debug/debug.go:
```diff
@@ -0,0 +1,53 @@
+// To see Info messages use a logger with `l.V(1).Enabled() == true`
+// To see Debug messages use a logger with `l.V(5).Enabled() == true`
+func SetLogger(l logr.Logger) {
+	globalLoggger = l
```
Because I removed the exported value this does have those guarantees.  Any call made after a SetLogger will be routed to the new logger, and because there is no way to capture this logger you don't need lazy evaluation or delegation.

MrAlias replied:
Gotcha :+1:.

There does still seem to be a race condition here where one of the logging functions is called concurrently with `SetLogger`.

MadVikingGod replied:
Can you explain more?  Because there is no check and then use we can't race here.  Yes if you call a log function before you call Set it will go to the old one, but there is no locking that would fix that.

MrAlias replied:
Maybe I'm missing something, but one goroutine calls `SetLogger` and another goroutine calls, say, `Info` since one is assigning to `globalLogger` and the other is reading it there is a race condition.

For example:

```
package debug

import (
	"log"
	"os"
	"testing"

	"github.com/go-logr/stdr"
)

func TestRace(t *testing.T) {
	go SetLogger(stdr.New(log.New(os.Stderr, "", 0)))
	go Info("")
}
```

and

```
go test -race
PASS
==================
WARNING: DATA RACE
Read at 0x0000006c2c50 by goroutine 9:
  runtime.racereadrange()
      <autogenerated>:1 +0x1b
  go.opentelemetry.io/otel/internal/debug.TestRace·dwrap·2()
      /home/tyler/go/src/go.opentelemetry.io/otel/internal/debug/debug_test.go:13 +0x71

Previous write at 0x0000006c2c50 by goroutine 8:
  go.opentelemetry.io/otel/internal/debug.SetLogger()
      /home/tyler/go/src/go.opentelemetry.io/otel/internal/debug/debug.go:36 +0x4b
  go.opentelemetry.io/otel/internal/debug.TestRace·dwrap·1()
      /home/tyler/go/src/go.opentelemetry.io/otel/internal/debug/debug_test.go:12 +0x58

Goroutine 9 (running) created at:
  go.opentelemetry.io/otel/internal/debug.TestRace()
      /home/tyler/go/src/go.opentelemetry.io/otel/internal/debug/debug_test.go:13 +0x26e
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/lib/go/src/testing/testing.go:1306 +0x47

Goroutine 8 (finished) created at:
  go.opentelemetry.io/otel/internal/debug.TestRace()
      /home/tyler/go/src/go.opentelemetry.io/otel/internal/debug/debug_test.go:12 +0x1cb
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/lib/go/src/testing/testing.go:1306 +0x47
==================
Found 1 data race(s)
exit status 66
FAIL	go.opentelemetry.io/otel/internal/debug	0.021s
```

MadVikingGod commented on internal/debug/debug.go:
```diff
@@ -0,0 +1,53 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package debug // import "go.opentelemetry.io/otel/internal/debug"
```
Just to be clear you are looking to move these methods to `/internal/`?  That would mean when we call these within our code we have `internal.Info(...)`, `internal.Error(err,...)` and `internal.Debug(...)`.

MrAlias replied:
I was talking about the [internal/global](https://github.com/open-telemetry/opentelemetry-go/tree/main/internal/global) package. I think flattening into that package (and we call `global.Info(...)`, `global.Error(err, ...)`, `global.Debug(...)`) would work, but it might also make sense to just have it become a subpackage as well.

pellared commented on internal/debug/debug.go:
```diff
@@ -0,0 +1,53 @@
+//
+// The default logger uses stdr which is backed by the standard `log.Logger`
+// interface. This logger will only show messages at the Error Level.
+var globalLoggger logr.Logger = stdr.New(log.New(os.Stderr, "", log.LstdFlags|log.Lshortfile))
```
Can we have a null (no-op) logger here?
As a rule of thumb, we should not affect the instrumented application in a "visible" way.

MadVikingGod replied:
This is done so we can enable error reporting, via ErrorHandeler. We should change that default first.

pellared replied:
> This is done so we can enable error reporting, via ErrorHandeler. 

Cannot the user enable it by himself via `otel.SetErrorHandler` ?

> We should change that default first.

I do not follow 😞 Can you rephrase

I think that it would be better just to include in this PR an internal API for logging. And not touch the ErrorHandler at all.

pellared replied:
Resolving per https://github.com/open-telemetry/opentelemetry-go/issues/2417

pellared commented on debug/doc.go:
````diff
@@ -0,0 +1,46 @@
+/*
+Package debug provides methods to control internal logging of opentelemetry.
+
+# Usage:
+The default logger will discard all messages, to see the logs the application
+developer needs to set a logger.
+
+```
+import (
+	"os"
+
+	"go.opentelemetry.io/otel/debug"
+	"github.com/go-logr/stdr"
+)
+
+func main() {
+	debug.SetLogger(
+		stdr.New(log.New(os.Stderr, "", log.LstdFlags|log.Lshortfile))
+	)
+
+...
+}
+```
````
Do not you prefer to create an "example" test? It is easier to keep the sample usage in sync with code,

pellared commented on CHANGELOG.md:
```diff
@@ -24,15 +24,17 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
-- Added an internal Logger.  This can be used by the SDK and API to provide users with feedback of the internal state. To enable verbose logs configure the logger which will print V(1) logs.  For debugging information confgure to print V(5) logs.
+- Added an internal Logger.
+  This can be used by the SDK and API to provide users with feedback of the internal state.
+  To enable verbose logs configure the logger which will print V(1) logs.  For debugging information configure to print V(5) logs. (#2343)
```
```suggestion
To enable verbose logs configure the logger which will print V(1) logs. For debugging information configure to print V(5) logs. (#2343)
```

pellared commented on CHANGELOG.md:
```diff
@@ -24,13 +24,17 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
   - The following struct types moved and are replaced with type aliases, since they are exposed to the user: `Observation`, `Measurement`.
   - The No-op implementations of sync and async instruments are no longer exported, new functions `sdkapi.NewNoopAsyncInstrument()` and `sdkapi.NewNoopSyncInstrument()` are provided instead. (#2271)
 - Update the SDK `BatchSpanProcessor` to export all queued spans when `ForceFlush` is called. (#2080, #2335)
+- The default ErrorHandler now uses the internal logger.  The default of logging to stderr is unchanged. (#2343)
```
This line can be removed.

MadVikingGod commented:
So I think I might not be expressing my concerns about why we need to changethe error handling well enough.  I made a quick demo of what the behavior is right now where we don't touch the error handler, [here](https://github.com/MadVikingGod/opentelemetry-go/tree/internal_logging_demo/example/logging-demo).  Running it you will see:
```
######## Default Logger #######
2021/11/24 18:32:49 Error from error handler
2021/11/24 18:32:49 debug.go:47: "msg"="logging error" "error"="Error from interal logger"
######## Setting New Logger #######
2021/11/24 18:32:49 Error from error handler
ERRO[0000] logging error                                 error="Error from interal logger"
INFO[0000] Logging from Info    ```

I think that it is surprising to change SetLogger, and then some messages are still using a different logger.  Especially as an application developer because I would see all of these logs as emitted from otel.

If we were to link them you get two key results:
1. By default errors across our ecosystem will have a common formatted.
2. You don't have to set logging behavior in two places.  Yes, the error handler is more powerful in you can do things other than log, but that functionality is kept by changing the default.

pellared commented:
> So I think I might not be expressing my concerns about why we need to change the error handling well enough. 

I see your concern. However, personally, I do not like that currently [the default ErrorHandler is logging](https://github.com/open-telemetry/opentelemetry-go/blob/56c743ba7998613f046f51954b7ccf0f911cd29f/handler.go#L69). As for me, the ErrorHandling logging concern could be addressed in a separate PR as a follow-up. I assume we are not planning a release in the following days.

Normally I would propose a PR that removes logging from the default error handler (that could be merged even before this PR). However, I do not time to work on that and participate in possible conversations 😞

MrAlias reviewed (approved):
Outside of the minor changelog suggestions this looks ready to merge.

I think this is a great launch point for our internal logging! Thanks for digging into this. :rocket:

pellared commented on sdk/trace/batch_span_processor.go:
```diff
@@ -236,6 +238,7 @@ func (bsp *batchSpanProcessor) exportSpans(ctx context.Context) error {
 	}
 
 	if l := len(bsp.batch); l > 0 {
+		global.Info("exporting spans", "count", len(bsp.batch))
```
Can we change this one to `Debug`. Otherwise, it gets noisy from start 😉

pellared commented:
I created an issue https://github.com/open-telemetry/opentelemetry-go/issues/2417 and resolved most of my comments.

XSAM commented on internal/global/internal_logging.go:
```diff
@@ -0,0 +1,63 @@
+	"github.com/go-logr/stdr"
+)
+
+// globalLogger is the logging interface used within the otel api and sdk provide deatails of the internals.
+//
+// The default logger uses stdr which is backed by the standard `log.Logger`
+// interface. This logger will only show messages at the Error Level.
+var globalLoggerLock = &sync.RWMutex{}
+var globalLogger logr.Logger = stdr.New(log.New(os.Stderr, "", log.LstdFlags|log.Lshortfile))
```
```suggestion
// globalLogger is the logging interface used within the otel api and sdk provide details of the internals.
//
// The default logger uses stdr which is backed by the standard `log.Logger`
// interface. This logger will only show messages at the Error Level.
var globalLogger logr.Logger = stdr.New(log.New(os.Stderr, "", log.LstdFlags|log.Lshortfile))
var globalLoggerLock = &sync.RWMutex{}
```

jmacd commented on internal/global/internal_logging.go:
```diff
@@ -0,0 +1,63 @@
+func Info(msg string, keysAndValues ...interface{}) {
+	globalLoggerLock.RLock()
+	defer globalLoggerLock.RUnlock()
+	globalLogger.V(1).Info(msg, keysAndValues...)
```
I'm curious about the choice of `V(1)` and `V(5)`. I take it the user shouldn't see OTel-SDK Info messages unless they're debugging, thus Info is `V(1)`?

MadVikingGod replied:
Yeah, the idea was that V(1) is akin to a command line `-v`, the extra data that might be helpful.
